webui/react/src/services/stream/stream.ts

```diff
@@ -0,0 +1,219 @@
+  #pendingMsg: Array<Record<string, StreamContent>> = [];
+
+  //callbacks
+  #onUpsert: (m: Record<string, StreamContent>) => void;
+  #onDelete: (s: Streamable, a: Array<number>) => void;
+  #isLoading: ((b: boolean) => void) | undefined;
+
+  constructor(
+    wsUrl: string,
+    onUpsert: (m: Record<string, StreamContent>) => void,
+    onDelete: (s: Streamable, a: Array<number>) => void,
+    isLoading?: (b: boolean) => void,
+  ) {
+    this.#wsUrl = wsUrl;
+    this.#onUpsert = onUpsert;
+    this.#onDelete = onDelete;
+    this.#isLoading = isLoading;
+    this.#advance();
+  }
+
+  #connect() {
+    this.#ws = new WebSocket(this.#wsUrl);
+    this.#ws.onopen = () => {
+      logger.info('Streaming websocket opened!');
+      this.#retries = 0;
+      this.#advance();
+    };
+    this.#ws.onerror = (err) => {
+      // No need to do anything else becauses onerror will trigger onclose
+      logger.error('Streaming websocket errored: ', err);
+    };
+
+    this.#ws.onclose = () => {
+      this.#syncSent = undefined;
+      const backoff = backoffs[this.#retries];
+      if (backoff === undefined) {
+        throw new Error('Websocket cannot reconnect!');
+      }
+      this.#timeout = dayjs().add(backoff, 'second');
+      this.#retries += 1;
+      logger.info(`#${this.#retries} of retries: in ${backoff}s`);
+      setTimeout(() => this.#advance(), backoff * 1000);
+    };
+
+    this.#ws.onmessage = (event) => {
+      const msg = JSON.parse(event.data) as Record<string, StreamContent>;
+      this.#pendingMsg.push(msg);
+      this.#advance();
+    };
+  }
+
+  #shouldSkip(newSub: Record<Streamable, Subscription>): boolean {
+    if (!this.#curSub) return false;
+    let same = true;
+    forEach(newSub, (val, k) => {
+      if (!this.#curSub?.[k as Streamable].spec.equals(val.spec)) same = false;
+    });
+    return same;
+  }
+
+  #sendSpec(newSub?: Record<Streamable, Subscription>): void {
+    if (!newSub) return;
+    // Skip current sub and move to next only if sub has already been sent
+    if (this.#shouldSkip(newSub) && this.#syncSent) {
+      this.#advance();
+      return;
+    }
+
+    this.#numSyncs += 1;
+    const sync_id = this.#numSyncs.toString();
+    const spec = reduce(
+      newSub,
+      (spec, ent, k) => {
+        spec['known'][k] = ent.keyCache.known();
+        spec['subscribe'][k] = {
+          ...ent.spec.toWire(),
+          since: ent.keyCache.maxSeq(),
+        };
+        return spec;
+      },
+      { known: {}, subscribe: {}, sync_id: sync_id } as Record<string, StreamContent>,
+    );
+
+    this.#curSub = newSub;
+    this.#ws && this.#ws.send(JSON.stringify(spec));
+    this.#syncSent = sync_id;
+  }
+
+  #processMsg(): void {
+    while (this.#pendingMsg.length > 0) {
+      const msg = this.#pendingMsg.shift();
+      if (!msg) break;
+      if (msg['sync_id']) {
+        if (!msg['complete']) {
+          this.#syncStarted = msg['sync_id'];
+          this.#isLoading?.(false);
```

Comment:
should `this.#isLoading?.(false)` be called as soon as we are in a not-sync-complete state?  So:
- on startup
- after each sync send

Otherwise, I'm concerned that it isn't actually a very useful signal, because when you send a new subscription, you'll need to wait for `isLoading(false)` so you know your initial payload has started a download and then wait for another `isLoading(true)` so you know your initial payload is down downloading.

oh hey, also I think `isLoading(false)` and `isLoading(true)` are reversed.  It would be correct if it was called `isLoaded()`.